gym/spaces/dict.py

```diff
@@ -49,28 +50,45 @@ def __init__(self, spaces=None, **spaces_kwargs):
                 space, Space
             ), "Values of the dict should be instances of gym.Space"
         super(Dict, self).__init__(
-            None, None
+            None, None, seed
         )  # None for shape and dtype, since it'll require special handling
 
     def seed(self, seed=None):
-        seed = super().seed(seed)
-        try:
-            subseeds = self.np_random.choice(
-                np.iinfo(int).max,
-                size=len(self.spaces),
-                replace=False,  # unique subseed for each subspace
-            )
-        except ValueError:
-            subseeds = self.np_random.choice(
-                np.iinfo(int).max,
-                size=len(self.spaces),
-                replace=True,  # we get more than INT_MAX subspaces
-            )
-
-        for subspace, subseed in zip(self.spaces.values(), subseeds):
-            seed.append(subspace.seed(int(subseed))[0])
-
-        return seed
+        seeds = []
+        if isinstance(seed, dict):
+            for key, seed_key in zip(self.spaces, seed):
+                assert key == seed_key, print(
+                    "Key value",
+                    seed_key,
+                    "in passed seed dict did not match key value",
+                    key,
+                    "in spaces Dict.",
+                )
+                seeds += self.spaces[key].seed(seed[seed_key])
+        elif isinstance(seed, int):
+            seeds = super().seed(seed)
+            try:
+                subseeds = self.np_random.choice(
+                    np.iinfo(int).max,
+                    size=len(self.spaces),
+                    replace=False,  # unique subseed for each subspace
+                )
+            except ValueError:
+                subseeds = self.np_random.choice(
+                    np.iinfo(int).max,
+                    size=len(self.spaces),
+                    replace=True,  # we get more than INT_MAX subspaces
+                )
+
+            for subspace, subseed in zip(self.spaces.values(), subseeds):
+                seeds.append(subspace.seed(int(subseed))[0])
+        elif seed is None:
+            for space in self.spaces.values():
+                seeds += space.seed(seed)
+        else:
+            raise TypeError("Passed seed not of an expected type: dict or int or None")
+
+        return seeds
```

Comment:
1. Should we only add the main seed of subspace (`subspace.seed()[0]`) to the return value rather than extend all items in `subspace.seed()` (the op `seeds +=` in the first and the last cases)?
  If the subpace of `Dict` here is another compound space (`Tuple` or `Dict`), the length of the return list could be different `len(d.seed(None)) != len(d.seed(0))`. (However, I think we probably only use the first item, and hardly ever use the rest in the list and the length of the list. This issue probably never affects normal users.)

2. We could merge the last two cases (`seed` is `int` or `None`) into one. After statement `seeds = super().seed(seed)`, `self.np_random` become seeded and variable `seeds` is a list of integers.

Reply:
1. I'd say it causes minimal harm and could be done. Doing this would create the `_np_random` PRNG of the "base" class's object and make it consistent with other spaces in case all `Space` objects are expected to have an `_np_random` that is not `None`, although I assume that only `_np_random`s of the sub-classes are ever used. 
2. It's subjective. For the case, when it's `None`, I assume the user would want the seed of every sub-space to be as random as possible in the sense that they would want all of them to be seeded with `None`. But generating the seeds for sub-spaces using an `_np_random` PRNG generated based on `None` (i.e., the suggested way) should also be fine.
3. I don't think that there's an expected fixed length of the list of seeds returned, so I do not think it makes a difference either way. In any case, the length of the list of seeds returned can always be calculated if the sub-spaces are known, in case a user really wants dig deeper.

So, in general, I'm fine with any of the suggestions either being or not being implemented (also for the suggestions on `Tuple` below). Please let me know.